File: baekjoon_step/13. 백트래킹/스도쿠.c
#include <stdio.h>
#include <stdlib.h>

int board[9][9];
int row[9][10];
int col[9][10];
int square[9][10];

void completeBoard(int level) {
	int x = level / 9, y = level % 9;
	if (level == 81) {
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++)
				printf("%d ", board[i][j]);
			printf("\n");
		}
		exit(0);
	}
	if (board[x][y] != 0) {
		completeBoard(level + 1);
		return;
	}
	for (int i = 1; i < 10; i++) {
		if (!row[x][i] && !col[y][i] && !square[3 * (x / 3) + y / 3][i]) {
			board[x][y] = i;
			row[x][i] = 1;
			col[y][i] = 1;
			square[3 * (x / 3) + y / 3][i] = 1;
			completeBoard(level + 1);
			board[x][y] = 0;
			row[x][i] = 0;
			col[y][i] = 0;
			square[3 * (x / 3) + y / 3][i] = 0;
		}
	}
}

int main() {
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			scanf("%d", &board[i][j]);
			row[i][board[i][j]] = 1;
			col[j][board[i][j]] = 1;
			square[3 * (i / 3) + j / 3][board[i][j]] = 1;
		}
		scanf("%*c");
	}

	completeBoard(0);

	return 0;
}

File: 백준/Gold/2239. 스도쿠/스도쿠.py
import sys


def back_tracking(level):
    global exit
    if level >= 81:
        for i in range(9):
            print("".join(map(str, row[i])))
        exit = True
        return
    if exit:
        return

    i = level // 9
    j = level % 9
    if matrix[i][j] != 0:
        back_tracking(level + 1)
    else:
        for num in range(1, 10):
            if num not in row[i] and num not in col[j] and num not in box[i // 3 * 3 + j // 3] and not exit:
                row[i][j] = num
                col[j][i] = num
                box[i // 3 * 3 + j // 3][i % 3 * 3 + j % 3] = num
                back_tracking(level + 1)
                row[i][j] = 0
                col[j][i] = 0
                box[i // 3 * 3 + j // 3][i % 3 * 3 + j % 3] = 0


matrix = []
row = [[] for _ in range(9)]
col = [[] for _ in range(9)]
box = [[] for _ in range(9)]

for i in range(9):
    temp = list(map(int, str(sys.stdin.readline().rstrip())))
    matrix.append(temp)
    for j in range(9):
        row[i].append(temp[j])
        col[j].append(temp[j])
        box[i // 3 * 3 + j // 3].append(temp[j])

exit = False
back_tracking(0)
